### PR TITLE
try 3

### PR DESCRIPTION
**Pull request by jaideep-seth**
#254

### CODE REVIEW

**Comment by jaideep-seth** [comment]
@mwatts15  I refered to 'query_context.py' while making evidence_for() method, please help me out with this!

1)  When using PyOpenWorm.connect('readme.conf') i get evidence for the corresponding Statements as shown below vs when using PyOpenWorm.connect(configFile='tests/data_integrity_test.conf') where i get no evidence, why is this?

>>> import PyOpenWorm as P
INFO:rdflib:RDFLib Version: 4.2.2
>>> conn = P.connect('readme.conf')
INFO:PyOpenWorm:Connected to database
>>> from PyOpenWorm.context import Context
>>> from PyOpenWorm.evidence import Evidence
>>> from PyOpenWorm.evidence import evidence_for
>>> from PyOpenWorm.neuron import Neuron
>>> qctx = Context()
>>> qctx(Neuron)('AVAL').innexin('UNC-7')
PyOpenWorm.statement.Statement(subj=Neuron(ident=rdflib.term.URIRef('http://openworm.org/entities/Neuron/AVAL')), prop=PyOpenWorm.neuron.Neuron_innexin(owner=Neuron(ident=rdflib.term.URIRef('http://openworm.org/entities/Neuron/AVAL'))), obj=yarom.propertyValue.PropertyValue(rdflib.term.Literal('UNC-7')), context=PyOpenWorm.context.Context())
>>> ev_objs = evidence_for(qctx, conn, Evidence, Context)
CONTEXT ZODBStore:DEFAULT
CONTEXT http://openworm.org/entities/data_sources/DataWithEvidenceDataSource#Neurons/context_for?document=http://openworm.org/entities/Document/Altun2009
CONTEXT http://openworm.org/entities/data_sources/DataWithEvidenceDataSource#Neurons/context_for?document=http://openworm.org/entities/Document/a7211bdf58d44f520ec6ffc0cda9dfc46
>>> ev_objs
[Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/ad2ad33a93b07157f9932a27059201a8c'))]



2) My PR - 'try 3', clears all the tests in my local virtual environment but is failing when i push it, i guess its due to a different PyOpenWorm.connect('') ? please help me address these issues.

P.S - i passed the Evidence and Context classes into evidence_for() as i wasnt being able to access them in PyOpenWorm.evidence

-Thank you!

**Comment by mwatts15** [comment]
Passing in the classes really should be unnecessary. I've tried it locally and it works.

As for why you see different behavior for the two different config files -- not sure: what directory are you running that code from and what version of Python are you using? I see the Travis-CI failure, but I wasn't able yet to reproduce the deviation between 3.6 and 2.7 locally.

**Comment by mwatts15** [comment]
I was able to reproduce in 3.7...

File: PyOpenWorm/evidence.py [hunk 1]
```diff
@@ -85,4 +85,37 @@ def identifier_augment(self):
...
+    """
+    Returns an iterable of Evidence  
+    """
+    ctxs = query_context(conn.conf['rdf.graph'], qctx)
+    for c in ctxs:
+        mqctx = Context(conf=conn.conf)
+        print('CONTEXT', c.identifier)
+        ev = mqctx.stored(Evidence)()
+        ev.supports(Context(ident=c.identifier, conf=conn.conf).rdf_object)
+        ev_objs = []
```
Comment (mwatts15) [inline PyOpenWorm/evidence.py]: this shouldn't be reset each time. It's just a fluke that it was different in different versions of Python -- the context which had nothing attached just happened to come last in the tests that failed.

File: tests/EvidenceForTest.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        xfail_without_db()
+        self.conn = PyOpenWorm.connect('readme.conf')
+        self.g = self.conn.conf["rdf.graph"]
+        self.context = Context()
+        self.qctx = self.context.stored
+
+    def tearDown(self):
+        PyOpenWorm.disconnect(self.conn)
+
+    def test_evidence_for(self):
```
Comment (mwatts15) [inline tests/EvidenceForTest.py]: It's OK to have this test, but you should *really* have self-contained tests. This one depends on the OpenWorm data, but we need to make sure `evidence_for` works.

File: PyOpenWorm/evidence.py (hunk 1 above)
Comment (jaideep-seth) [reply PyOpenWorm/evidence.py]: Okay, got it - [Document, Document, None]

**Comment by jaideep-seth** [comment]
hey, @mwatts15 

-The last commit has self-contained tests and is pep-8 compliant...please tell me what else can i do to get this merged! 

-Thank you for the guidance

**Comment by jaideep-seth** [comment]
illustrated in Python 3.6 virtual environment:

>>> import PyOpenWorm as P
INFO:rdflib:RDFLib Version: 4.2.2
>>> conn = P.connect('readme.conf')
INFO:PyOpenWorm:Connected to database
>>> from PyOpenWorm.context import Context
>>> c1 = Context()
>>> from PyOpenWorm.evidence import Evidence
>>> from PyOpenWorm.evidence import evidence_for
>>> from PyOpenWorm.neuron import Neuron
>>> c1(Neuron)('AVAL').innexin('UNC-7')
PyOpenWorm.statement.Statement(subj=Neuron(ident=rdflib.term.URIRef('http://openworm.org/entities/Neuron/AVAL')), prop=PyOpenWorm.neuron.Neuron_innexin(owner=Neuron(ident=rdflib.term.URIRef('http://openworm.org/entities/Neuron/AVAL'))), obj=yarom.propertyValue.PropertyValue(rdflib.term.Literal('UNC-7')), context=PyOpenWorm.context.Context())
>>> evc = Context()
>>> ev1 = evc(Evidence)(key='js2019')
>>> ev1.supports(c1.rdf_object)
PyOpenWorm.statement.Statement(subj=Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/js2019')), prop=PyOpenWorm.evidence.Evidence_supports(owner=Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/js2019'))), obj=ContextDataObject(ident=rdflib.term.Variable('aContextDataObject_1')), context=PyOpenWorm.context.Context())
>>> ctx = Context()
>>> ctx.add_import(c1)
>>> ctx.add_import(evc)
>>> ctx.save_context()
>>> ev_iterable = evidence_for(c1, conn)
CONTEXT ZODBStore:DEFAULT
CONTEXT http://openworm.org/entities/data_sources/DataWithEvidenceDataSource#Neurons/context_for?document=http://openworm.org/entities/Document/Altun2009
CONTEXT http://openworm.org/entities/data_sources/DataWithEvidenceDataSource#Neurons/context_for?document=http://openworm.org/entities/Document/a7211bdf58d44f520ec6ffc0cda9dfc46
>>> ev_iterable
[Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/aca86d8acbde8e6e0ea63cbae9276ab7f')), Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/ad2ad33a93b07157f9932a27059201a8c'))]
>>>

File: tests/EvidenceForTest.py [hunk 3]
```diff
@@ -28,7 +27,14 @@ def tearDown(self):
         PyOpenWorm.disconnect(self.conn)
 
     def test_evidence_for(self):
-        qctx = Context()
-        qctx(Neuron)('AVAL').innexin('UNC-7')
-        ev_iterable = evidence_for(qctx, self.conn, Evidence, Context)
-        self.assertTrue((len(ev_iterable) != 0))
\ No newline at end of file
+        c1 = Context()
```
Comment (mwatts15) [inline tests/EvidenceForTest.py]: Not quite -- please have this TestCase subclass DataTestTemplate._DataTest and you'll see this test still fails.

**Comment by jaideep-seth** [comment]
hey, @mwatts15 

1) I used https://media.readthedocs.org/pdf/pyopenworm/latest/pyopenworm.pdf#section.2.4 to make my Apple : 
>>> class Apple(DataObject):
...     def __init__(self, color=False, *args, **kwargs):
...             DataObject.__init__(self, *args, **kwargs)
...             Apple.DatatypeProperty('sweetness', owner=self)
...             Apple.DatatypeProperty('color', owner=self)
...             if color:
...                     self.color(color)

Then i made a statement about an Apple:
>>> ctx(Apple)('green').sweetness('medium')
PyOpenWorm.statement.Statement(subj=Apple(ident=rdflib.term.Variable('aApple_1')), prop=__main__.Apple_sweetness(owner=Apple(ident=rdflib.term.Variable('aApple_1'))), obj=yarom.propertyValue.PropertyValue(rdflib.term.Literal('medium')), context=PyOpenWorm.context.Context())

I looked into PyOpenWorm/examples/add_reference.py to add evidence for the statement, ctx(Apple)('green').sweetness('medium'), and saved it as:

>>> evctx = Context(ident='http://example.org/meta', conf=conn.conf)
>>> e = evctx(Evidence)(key='js2019')
>>> e.supports(ctx.rdf_object)
PyOpenWorm.statement.Statement(subj=Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/js2019')), prop=PyOpenWorm.evidence.Evidence_supports(owner=Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/js2019'))), obj=ContextDataObject(ident=rdflib.term.Variable('aContextDataObject_1')), context=PyOpenWorm.context.Context(ident="http://example.org/meta"))
>>> ctx.save_context()
>>> evctx.save_context()

and viewed the evidence as:
>>> for e_i in evctx.stored(Evidence)().load():
...     print(e_i.reference(), e_i.supports())
... 
Document(Document:Anonymous2011) DataObject(<http://example.org/data>)
None DataObject(<http://example.org/data>)
None DataObject(<http://example.org/data>)


2) I referred to PyOpenWorm/examples/query_context.py to create `evidence_for()`, It returns an iterable of Evidence objects for the statement, ctx(Neuron)('AVAL').innexin('UNC-7'), as below:
>>> ctx(Neuron)('AVAL').innexin('UNC-7')
PyOpenWorm.statement.Statement(subj=Neuron(ident=rdflib.term.URIRef('http://openworm.org/entities/Neuron/AVAL')), prop=PyOpenWorm.neuron.Neuron_innexin(owner=Neuron(ident=rdflib.term.URIRef('http://openworm.org/entities/Neuron/AVAL'))), obj=yarom.propertyValue.PropertyValue(rdflib.term.Literal('UNC-7')), context=PyOpenWorm.context.Context())
>>> from PyOpenWorm.evidence import evidence_for
>>> evidence_for(ctx, conn)
[Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/ad2ad33a93b07157f9932a27059201a8c')), Evidence(ident=rdflib.term.URIRef('http://openworm.org/entities/Evidence/aca86d8acbde8e6e0ea63cbae9276ab7f'))]

but `evidence_for` returns an empty list for the statement, ctx(Apple)('green').sweetness('medium'), as:
>>> ctx = Context()
>>> ctx(Apple)('green').sweetness('medium')
PyOpenWorm.statement.Statement(subj=Apple(ident=rdflib.term.Variable('aApple_1')), prop=__main__.Apple_sweetness(owner=Apple(ident=rdflib.term.Variable('aApple_1'))), obj=yarom.propertyValue.PropertyValue(rdflib.term.Literal('medium')), context=PyOpenWorm.context.Context())
>>> evidence_for(ctx, conn)
[]

Why is this, i just added evidence for this statement in point 1?

3) How can i save my Apple and its evidence into |pow| such that it too can be retrieved in the same manner as ctx(Neuron)('AVAL').innexin('UNC-7')


-Thank you!

**Comment by mwatts15** [comment]
@jaideep-seth notice that you have not given an identifier for the `Apple` you created. Generally, the identifier will be generated from a property or set of properties. Have a look at some of the types which have already been implemented.

File: tests/EvidenceForTest.py (hunk 3 above)
Comment (mwatts15) [reply tests/EvidenceForTest.py]: Sorry: also remove the setUp/tearDown implementations -- the point is that you are using the OpenWormData database, but should not be.

**Comment by jaideep-seth** [comment]
I will try and further continue on adding support for querying for evidence for DataObjects rather than Contexts and for querying for Contexts that do have an identifier!

**Pull request merged**